### ISSUE TITLE
Tailscale 1.50 | `serve` config purged after update

### ISSUE THREAD

**Issue by gromoslaw-kroczka** [tailscale/tailscale#9558]
### What is the issue?

After update of Docker Image to v1.50 `serve` (reverse proxy) and funnel settings are purged after restart of the Tailscale.

Have to rolled back to 1.48.2 to be able to reboot the container without losing serve settings.

### Steps to reproduce

1) Update to image v1.50 on Docker Container.
2) Previous `serve` settings will be purged
3) After assigned new `serve` rules and restart of the container data will be purged again

### Are there any recent changes that introduced the issue?

For me it is only present on v1.50. After rollback to v1.48.2 everything work smooth (config preserved after restarts).

### OS

Linux

### OS version

Debian 11

### Tailscale version

1.50.0

**Comment by guest42069** [comment]
Also experiencing this.

Set up serve for a web service, check service and check the tailscaled.state file for the `_serve` config.

```
[root@rpi ~]# tailscale version
1.50.0
  go version: go1.21.1
[root@rpi ~]# tailscale serve status
No serve config
[root@rpi ~]# tailscale serve https /jellyfin http://127.0.0.1:8096
[root@rpi ~]# tailscale serve status
https://rpi.[snip].ts.net (tailnet only)
|-- /jellyfin proxy http://127.0.0.1:8096

[root@rpi ~]# curl -I https://rpi.[snip].ts.net/jellyfin
HTTP/2 302 
date: Thu, 28 Sep 2023 17:44:35 GMT
location: /jellyfin/web/index.html
server: Kestrel

[root@rpi ~]# grep "_serve" `podman volume inspect -f '{{.Mountpoint}}' tailscaled-state`/tailscaled.state
  "_serve/d948": "eyJUQ1AiOnsiNDQzIjp7IkhUVFBTIjp0cnVlfX0sIldlYiI6eyJycGkuW3NuaXBdLnRzLm5ldDo0NDMiOnsiSGFuZGxlcnMiOnsiL2plbGx5ZmluIjp7IlByb3h5IjoiaHR0cDovLzEyNy4wLjAuMTo4MDk2In19fX19",
```
serve is working as expected and the entry exists in the config stored in the tailscaled.state file

```
[root@rpi ~]# systemctl restart container-tailscaled
```
restart the service (container)

```
[root@rpi ~]# tailscale serve status
No serve config
[root@rpi ~]# curl -I https://rpi.[snip].ts.net/jellyfin
curl: (7) Failed to connect to rpi.[snip].ts.net port 443 after 7220 ms: Couldn't connect to server
[root@rpi ~]# grep "_serve" `podman volume inspect -f '{{.Mountpoint}}' tailscaled-state`/tailscaled.state 
  "_serve/d948": "e30=",
```
the serve config is no longer functional and the entry in tailscaled.state is now blank.

---

I did some digging and was able to find a work-around for my use case, but I'm not sure if this is intended.

My tailscale runs inside a container, and I notice there is also via containerboot a `TS_SERVE_CONFIG` environment variable which doesn't have much documentation, but I found it's basically a path to a json encoded `ipn.ServeConfig`. The base64 encoded blob associated with the `_serve` entry in the tailscaled.state file is the same.

So I created a file with the decoded _serve entry json, gave the container access and set the `TS_SERVE_CONFIG` to the path to the file.

```
[root@rpi ~]# podman exec tailscaled sh -c 'cat $TS_SERVE_CONFIG'
{"TCP":{"443":{"HTTPS":true},"8443":{"HTTPS":true}},"Web":{"rpi.[snip].ts.net:443":{"Handlers":{"/grafana":{"Proxy":"http://127.0.0.1:3000"},"/jellyfin":{"Proxy":"http://127.0.0.1:8096"},"/prometheus":{"Proxy":"http://127.0.0.1:9090"}}},"rpi.[snip].ts.net:8443":{"Handlers":{"/":{"Proxy":"http://127.0.0.1:9100"}}}}}
[root@rpi ~]# tailscale serve status
https://rpi.[snip].ts.net (tailnet only)
|-- /grafana    proxy http://127.0.0.1:3000
|-- /jellyfin   proxy http://127.0.0.1:8096
|-- /prometheus proxy http://127.0.0.1:9090

https://rpi.[snip].ts.net:8443 (tailnet only)
|-- / proxy http://127.0.0.1:9100

```

**Comment by DentonGentry** [comment]
Unstable build 1.51.30 contains a fix for this, the serve config will be retained unless explicitly changed.
If able to run an unstable release, could you try it?

**Comment by DentonGentry** [comment]
This should be resolved in 1.50.1, released earlier today.
